### PR TITLE
CLN: Clean up some iteration logic in tslib

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by lithomas1** [comment]
I don't think the multi-iterator is necessary, we only iterate on values not results.

cc @jbrockmendel

File: pandas/_libs/tslib.pyx [hunk 1]
```diff
@@ -455,18 +455,18 @@ cpdef array_to_datetime(
...
     assert is_raise or is_ignore or is_coerce
 
     result = np.empty((<object>values).shape, dtype="M8[ns]")
-    mi = cnp.PyArray_MultiIterNew2(result, values)
     iresult = result.view("i8").ravel()
 
     for i in range(n):
         # Analogous to `val = values[i]`
-        val = <object>(<PyObject**>cnp.PyArray_MultiIter_DATA(mi, 1))[0]
+        val = cnp.PyArray_GETITEM(values, cnp.PyArray_ITER_DATA(it))
```
Comment (jbrockmendel) [inline pandas/_libs/tslib.pyx:468]: elsewhere we use `val = (<int64_t*>cnp.PyArray_ITER_DATA(it))[0]`.  should we use PyArray_GETITEM there too?  seems cleaner

Comment (lithomas1) [reply pandas/_libs/tslib.pyx:468]: This returns a PyObject I think, so I don't think we can use it unless the array is object dtype.

File: pandas/_libs/tslib.pyx [hunk 2]
```diff
@@ -455,18 +455,18 @@ cpdef array_to_datetime(
         set out_tzoffset_vals = set()
         tzinfo tz_out = None
         bint found_tz = False, found_naive = False
-        cnp.broadcast mi
+        cnp.flatiter it = cnp.PyArray_IterNew(values)
```
Comment (WillAyd) [inline pandas/_libs/tslib.pyx:458]: I think PyArray_IterNew is the old function and we should be using NpyIter_New instead - does Cython offer that?

https://numpy.org/doc/stable/reference/c-api/iterator.html#converting-from-previous-numpy-iterators

File: pandas/_libs/tslib.pyx [hunk 3]
```diff
@@ -455,18 +455,18 @@ cpdef array_to_datetime(
...
 
     result = np.empty((<object>values).shape, dtype="M8[ns]")
-    mi = cnp.PyArray_MultiIterNew2(result, values)
     iresult = result.view("i8").ravel()
 
     for i in range(n):
         # Analogous to `val = values[i]`
-        val = <object>(<PyObject**>cnp.PyArray_MultiIter_DATA(mi, 1))[0]
+        val = cnp.PyArray_GETITEM(values, cnp.PyArray_ITER_DATA(it))
+        cnp.PyArray_ITER_NEXT(it)
```
Comment (WillAyd) [inline pandas/_libs/tslib.pyx:469]: Similar to comment above but worth closer inspection we should be using the iteration function https://numpy.org/doc/stable/reference/c-api/iterator.html#functions-for-iteration

File: pandas/_libs/tslib.pyx (hunk 2 above)
Comment (lithomas1) [reply pandas/_libs/tslib.pyx:458]: It'd doesn't seem to be exposed, there's an open issue :( . 

https://github.com/numpy/numpy/issues/21014.

**Review by WillAyd — APPROVED** [review]

**Review by jbrockmendel — APPROVED** [review]

**Comment by jbrockmendel** [comment]
thanks @lithomas1

**Pull request merged**